src/documentation/programming-guide.md

```diff
@@ -1283,7 +1283,225 @@ public class MyCustomDataType {
+
+To use windowing with fixed data sets, you can assign your own timestamps to each element. To assign timestamps to elements, use a `ParDo` transform with a `DoFn` that outputs each element with a new timestamp (for example, the [WithTimestamps]({{ site.baseurl }}/documentation/sdks/javadoc/{{ site.release_latest }}/index.html?org/apache/beam/sdk/transforms/WithTimestamps.html) transform in the Beam SDK for Java).
+
+To illustrate how windowing with a bounded `PCollection` can affect how your pipeline processes data, consider the following pipeline:
+
+![Diagram of GroupByKey and ParDo without windowing, on a bounded collection]({{ "/images/unwindowed-pipeline-bounded.png" | prepend: site.baseurl }} "GroupByKey and ParDo without windowing, on a bounded collection")
+
+**Figure:** `GroupByKey` and `ParDo` without windowing, on a bounded collection.
+
+In the above pipeline, we create a bounded `PCollection` by reading a set of key/value pairs using `TextIO`. We then group the collection using `GroupByKey`, and apply a `ParDo` transform to the grouped `PCollection`. In this example, the `GroupByKey` creates a collection of unique keys, and then `ParDo` gets applied exactly once per key.
+
+Note that even if you don’t set a windowing function, there is still a window -- all elements in your `PCollection` are assigned to a single global window.
+
+Now, consider the same pipeline, but using a windowing function:
+
+![Diagram of GroupByKey and ParDo with windowing, on a bounded collection]({{ "/images/windowing-pipeline-bounded.png" | prepend: site.baseurl }} "GroupByKey and ParDo with windowing, on a bounded collection")
+
+**Figure:** `GroupByKey` and `ParDo` with windowing, on a bounded collection.
+
+As before, the pipeline creates a bounded `PCollection` of key/value pairs. We then set a [windowing function](#setwindowingfunction) for that `PCollection`. The `GroupByKey` transform groups the elements of the `PCollection` by both key and window, based on the windowing function. The subsequent `ParDo` transform gets applied multiple times per key, once for each window.
+
+### Windowing functions
+
+You can define different kinds of windows to divide the elements of your `PCollection`. Beam provides several windowing functions, including:
+
+*  Fixed Time Windows
+*  Sliding Time Windows
+*  Per-Session Windows
+*  Single Global Window
+*  Calendar-based Windows (not supported by the Beam SDK for Python)
+
+Note that each element can logically belong to more than one window, depending on the windowing function you use. Sliding time windowing, for example, creates overlapping windows wherein a single element can be assigned to multiple windows.
+
+#### Fixed time windows
+
+The simplest form of windowing is using **fixed time windows**: given a timestamped `PCollection` which might be continuously updating, each window might capture (for example) five minutes worth of elements.
```

Comment:
Somewhere around here, in case the user has paged out the processing time vs. event time distinction, it might be worth recalling it. Particularly `five minutes worth of elements` sounds a bit like processing time, versus something like `might capture all elements with timestamps that fall into a five minute interval`.